asdf/asdf.py

```diff
@@ -1755,6 +1758,13 @@ def is_asdf_file(fd):
         # If it's an InputStream let ASDF deal with it.
         return True
 
+    if isinstance(fd, str):
```

Comment:
Is it reasonable to detect ASDF vs ASDF-in-FITS using the URL's filename extension?  Ideally we would peek the file content here too, but doing that requires making multiple HTTP requests or passing a GenericFile object to astropy.io.fits.  astropy can handle GenericFile but doesn't treat it as a real file and I'm not sure what the consequences of that are.  Passing the URL to astropy also brings the benefit of cached downloads.

Reply:
You mean something like http://somewhere.edu/somepath/data.fits[asdfextention]? (I don't actually know the right syntax for identifying the extension; by number only, extname, extname+extver, etc.)

Reply:
I just mean `filename.fits` vs `filename.asdf`.  The implication is that the library won't be able to open an obfuscated URL that serves ASDF, like `https://example.com/files/441ec2c4-a18c-42b5-b6b4-9bc92cfeb013`.

Reply:
Another option is to check for `.fits`, `.fits.gz`, `.asdf`, and use that if present but download the first few bytes of the file if not.

Reply:
I'm not sure I understand the need to support asdf in fits implicitly. Is there a good use case?

Reply:
Initially, I'd be happy with requiring the ".asdf" extension. We can always expand the possibilities later.

Reply:
> I'm not sure I understand the need to support asdf in fits implicitly. Is there a good use case?

The initial bug report was from a specutils user opening an ASDF-in-FITS over https.  Basically they're doing this:

```python
asdf.open("https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/IFU_optimal_extraction/NRS00001-faintQSO-F100LP-G140H-01_1_491_SE_2020-08-25T12h15m00_s3d.fits")
```

This `is_asdf_file` function determines whether we open the file as ASDF or as FITS.  It's not the only check -- if we proceed with opening as ASDF, there will still be an error if the initial bytes are unrecognized, but at that point we will have lost our opportunity to open as FITS.

Reply:
I think this should work with `AsdfInFits` if possible, as that's our current use case.  Not sure the best implementation.  But yeah, I think if you're passing a `.fits` file to `asdf.open()`, then the assumption that it is an `AsdfInFits` file should be valid.

Reply:
Should we make ASDF the default?  The way I've coded it the library will open the file as FITS unless there's a `.asdf` extension present.  Might be more fitting to turn that around so that FITS only happens if the URL has a `.fits` or `.fits.gz` extension and assume everything else is ASDF.

Reply:
Why not require the http string include some information for ASDF to know that it is ASDF in FITS? That would be much simpler to implement.

Reply:
I don't think it's a good idea to require a URL string to inform `asdf` whether it is a good file to open or not.  We don't have any control in how Box formats its URLs.  And it adds extra complexity for the user if they are linking to data on Box.  It adds an extra failure point that can very easily be dealt with in the `asdf` code.

I think if the user is trying to open a fits file with `asdf.open()`, the assumption should be it is `AsdfInFits`. If it doesn't work, a useful error is raised.

Now how we determine it's a FITS file could be `.fits` or `.fits.gz` extension, or some bytes at the beginning.  I don't have a preference.  But perhaps going the simple route of extension is the best for now.  In the end, we want this to work for URLs of JWST .fits data.

Agree that opening as `.asdf` should be the default and only trigger `AsdfInFits` if there's an extension indicating.

Reply:
We could move towards requiring ADSF-in-FITS users to open with `asdf.fits_embed.open` or something like that.  I'd prefer that over the current strategy of auto-detecting with "one open method to rule them all".  It would be nice too to keep the FITS support quarantined to that particular module instead of distributing it around the library.  Maybe someday we'll want to move it into a separate package.

Reply:
It all depends on where this URL comes from doesn't it? Could I see an example of a blind URL (i.e., just copying and pasting one into an open function without any knowledge of the context)? If you know the context, you can explicitly provide that to the open function. Hey, this is an ASDF in FITS case, open it as such (using a keyword, or special open function). For a file on the local file system I can see trying to figure out what it is since that is low cost. But a remote http request is less so. You might as well copy the whole file and then do that.

Reply:
That could work, but that would involve some upstream breakage of existing code in `specutils`, `jwst`, `romancal`, etc., as currently datamodels, and most other packages use `asdf.open()`, and there's no reason for a package to use separate open function if it's going to be a URL, because a URL or a file are only known at runtime.  So it would always have to use `asdf.fits_embed.open` if it knew it was `AsdfInFits`.

Reply:
The "one open method" solution would look something like this:

- Request the URL and wait for the first 5 bytes to arrive.  Peek their values without consuming any of the stream.
- If `#ASDF`, pass on to AsdfFile
- If anything else, pass on to AsdfInFits

After digging around yesterday I'm of the opinion that this isn't feasible under the current structure of asdf.py (that module is a superfund site!), and it may not be wise to include a dramatic refactor in a 2.7.x release.

Reply:
(the ideal solution I mean, but if we want to get there then a band-aid based on the URL filename extension seems like a good short-term solution)

Reply:
Agree that the bandaid might be best for now, as we should not be doing breaking changes in a 2.7.4 release.  So for 2.7.x (and 2.8x I would argue), `asdf.open()` should open `AsdfInFits` files in the same way it always has.

Agree we should revisit this when we get a chance to refactor. I see there's already a #TODO in there already to fix `_open_impl`

https://github.com/asdf-format/asdf/blob/ba09e57d9344206016127d1a86d3319a2df9f2ec/asdf/asdf.py#L925-L929

Reply:
I kept the URL-based detection but made ASDF the default, and added a helpful warning when the library is forced to assume ASDF.